File: c++/array/JumpGame.cpp
class Solution {
public:
    bool canJump(int A[], int n) {
        if (n == 0 || n == 1) {
            return true;
        }
        int maxCanReach = A[0];
        int i = 1;
        while (i < maxCanReach + 1) {
            maxCanReach = std::max(A[i] + i, maxCanReach);
            if (maxCanReach >= n - 1) {
                return true;
            }
            i ++;
        }
        return false;
    }
};

File: python/array/JumpGame.py
class Solution:
    # @param A, a list of integers
    # @return a boolean
    def canJump(self, A):
        if len(A) == 0 or len(A) == 1:
            return True
            
        i = 1
        maxCanReach = A[0]
        while i < maxCanReach + 1:
            maxCanReach = max(A[i]+i, maxCanReach)
            if maxCanReach >= len(A) - 1:
                return True
            i += 1

        return False


if __name__ == "__main__":
    solution = Solution()

    print solution.canJump([1, 1, 1, 0])
